### PR TITLE
Faster renderables update

### PR DESCRIPTION
**Pull request by AndrewMeadows**
* optimize low-hanging fruit in EntityTreeRenderer::update() context

### CODE REVIEW

**Comment by AndrewMeadows** [comment]
This PR is an optimization pass over the code executed in `EntityTreeRenderer::update()`.

**Test Plan**

**(1)** Login to dev-welcome and find a spot off to the side that looks toward most of the objects in the domain.

**(2)** Logout to save your login location.  This location will be used to compare performance stats, so don't change it or the avatar's look-at direction for the initial comparison tests:

**(3)** Login to your known location using the latest dev-download interface client.

**(4)** Make sure these timing options are set as such:
**Developer --> Timing --> Performance Timing --> Show Timing Details** = enabled
**Developer --> Timing --> Performance Timing --> Show Top Ten** = disabled
**Developer --> Timing --> Performance Timing --> Expand /update** = enabled
**Developer --> Timing --> Performance Timing --> Expand /physics** = enabled

**(5)** Press `/` key to toggle the stats visible.

**(6)** Click the mouse on the stats to toggle the details visible.

**(7)** Pay attention to the stats value for:  `/idle/update/physics/harvestChanges`
The value may slowly oscillate over several seconds, but pay attention to the lowest value it shows.

**(8)** Logout, then back in to the same spot and repeat measure in (7) to make sure the approximate value is repeatable.

**(9)** Repeat steps (1) through (8) using the interface client from this PR.
--> The time spent in the `/idle/update/physics/harvestChanges` context should be smaller when using this PR.

**(10)** Walk around dev-welcome looking for objects that are animated (there's a flag blowing in the wind and other objects).
--> Verify animated models are still animating in this PR.

**(11)** Throw some dynamic objects around
--> verify dynamic objects still appear to move

**(12)** Find another avatar.
--> verify other avatars still render and update correctly

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4790,11 +4789,11 @@ void Application::update(float deltaTime) {
     QSharedPointer<AvatarManager> avatarManager = DependencyManager::get<AvatarManager>();
 
     {
-        PROFILE_RANGE_EX(simulation_physics, "Physics", 0xffff0000, (uint64_t)getActiveDisplayPlugin()->presentCount());
+        PROFILE_RANGE(simulation_physics, "Physics");
```
Comment (AndrewMeadows) [inline interface/src/Application.cpp]: The extra information being stored in some of these stats was not helpful, so I removed them.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4196/HighFidelity-Beta-PR11611-102a1cb36beee5faa15ee4230fa8cafcc683010b.dmg

File: libraries/entities-renderer/src/EntityTreeRenderer.cpp [hunk 2]
```diff
@@ -232,29 +237,22 @@ void EntityTreeRenderer::updateChangedEntities(const render::ScenePointer& scene
...
-                    PerformanceTimer pt("change");
-                    updateChangedEntities(scene, transaction);
-                }
-                {
-                    PerformanceTimer pt("enqueue");
-                    scene->enqueueTransaction(transaction);
-                }
+                addPendingEntities(scene, transaction);
+                updateChangedEntities(scene, transaction);
+                scene->enqueueTransaction(transaction);
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/EntityTreeRenderer.cpp]: I moved the `PerformanceTimer` instances to be inside the function calls themselves.  Except for the `enqueueTransaction()` method -- that call is so cheap it doesn't make sense to use a `PerformanceTimer` on it.

File: libraries/entities-renderer/src/RenderableEntityItem.cpp [hunk 3]
```diff
@@ -271,6 +274,9 @@ void EntityRenderer::removeFromScene(const ScenePointer& scene, Transaction& tra
 }
 
 void EntityRenderer::updateInScene(const ScenePointer& scene, Transaction& transaction) {
+#ifdef WANT_DETAILED_PROFILING
+    PROFILE_RANGE(simulation_physics, __FUNCTION__);
+#endif
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableEntityItem.cpp]: These profiling contexts were useful for debugging, but they are called **a lot** and their performance impact can be measured in busy domains.   I is not helpful to exposed them by default in release builds.

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 4]
```diff
@@ -151,7 +160,7 @@ bool RenderableModelEntityItem::needsUpdateModelBounds() const {
         return true;
     }
 
-    if (isMovingRelativeToParent() || isAnimatingSomething()) {
+    if (isAnimatingSomething()) {
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: This `isMovingRelativeToParent()` check was wasteful.  The check for changing `transform` below suffices to catch this case.

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 5]
```diff
@@ -178,13 +187,63 @@ bool RenderableModelEntityItem::needsUpdateModelBounds() const {
...
+        // TODO: there is opportunity to further optimize the isAnimatingSomething() case.
+        model->simulate(0.0f);
+        _needsInitialSimulation = false;
         _needsJointSimulation = false;
+        updateRenderItems = true;
+    }
+
+    if (updateRenderItems) {
+        model->updateRenderItems();
     }
```
Comment (AndrewMeadows) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: Rather than check ALL possible reasons for updating the model bounds and then doing ALL of the work, we now check for the various cases and try to update the cases that have changed.

And we only `updateRenderItems()` when something has actually changed.

File: libraries/entities/src/ModelEntityItem.cpp [hunk 6]
```diff
@@ -557,12 +557,6 @@ void ModelEntityItem::setAnimationLoop(bool loop) {
     });
 }
 
-bool ModelEntityItem::getAnimationLoop() const { 
-    return resultWithReadLock<bool>([&] {
-        return _animationProperties.getLoop();
-    });
-}
```
Comment (AndrewMeadows) [inline libraries/entities/src/ModelEntityItem.cpp]: Unused!

File: libraries/render-utils/src/Model.cpp [hunk 7]
```diff
@@ -133,6 +131,14 @@ void Model::setRotation(const glm::quat& rotation) {
...
 }
 
+// temporary HACK: set transform while avoiding implicit calls to updateRenderItems()
+// TODO: make setRotation() and friends set flag to be used later to decide to updateRenderItems()
+void Model::setTransformNoUpdateRenderItems(const Transform& transform) {
+    _translation = transform.getTranslation();
+    _rotation = transform.getRotation();
+    // DO NOT call updateRenderItems() here!
+}
+
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/Model.cpp]: There are more optimizations that can be made here, but we'll do them in a different PR.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4199/HighFidelity-Beta-PR11611-1f09c10fb15efb5c89f7bf04978eb1174b924fcf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4199/HighFidelity-Beta-PR11611-1f09c10fb15efb5c89f7bf04978eb1174b924fcf.exe

File: libraries/animation/src/Rig.cpp [hunk 8]
```diff
@@ -999,14 +1002,17 @@ void Rig::updateAnimationStateHandlers() { // called on avatar update thread (wh
...
     PROFILE_RANGE_EX(simulation_animation_detail, __FUNCTION__, 0xffff00ff, 0);
     PerformanceTimer perfTimer("updateAnimations");
+#endif
 
     setModelOffset(rootTransform);
 
     if (_animNode && _enabledAnimations) {
+#ifdef WANT_DETAILED_PROFILING
         PerformanceTimer perfTimer("handleTriggers");
+#endif
```
Comment (AndrewMeadows) [inline libraries/animation/src/Rig.cpp]: Per-frame profiling is ok, but not per-frame-per-object profiling.  The impact on the release build, with lots of objects in view, is measurable.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4203/HighFidelity-Beta-PR11611-339edc67c8efca481754b0147616be527e47ed91.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4203/HighFidelity-Beta-PR11611-339edc67c8efca481754b0147616be527e47ed91.exe

File: libraries/entities-renderer/src/RenderableModelEntityItem.cpp [hunk 9]
```diff
@@ -107,6 +110,9 @@ QVariantMap parseTexturesToMap(QString textures, const QVariantMap& defaultTextu
 }
 
 void RenderableModelEntityItem::doInitialModelSimulation() {
+#ifdef WANT_DETAILED_PROFILING
+    PROFILE_RANGE(simulation_physics, __FUNCTION__);
```
Comment (jherico) [inline libraries/entities-renderer/src/RenderableModelEntityItem.cpp]: you might consider adding a tracing category of `simulation_physics_detail` and migrating these to that instead.  By default most of the tracing scripts exclude any logging category that ends in `.detail`.

**Comment by AndrewMeadows** [comment]
Hrm... I just tried to follow my own test plan and found it rather hard to measure performance changes.  The stat measured for harvestChanges varies over time and so it is difficult to compare one to the other.  A simple range measurement for my test case shows the time there changing from [3.9, 4.9] msec to [3.1, 4.1] msec, which isn't much.

Another test where I was keeping several dynamic objects active in dev-welcome (the castle of blocks) and meausuring averages of chrome profiling stats showed:

5.264 msec average spent in "HarvestChanges" (in hifi/master) going to 4.445 msec average in "PostStep" (in this PR).

In short: performance gains in this PR are very small.

**Comment by Nex-Pro** [comment]
@AndrewMeadows I read your last comment, should this be QA'd based on the existing test plan ?

**Comment by AndrewMeadows** [comment]
I'm going to look into removing the extra detail profiling changes in this PR, or maybe try using the "detail" option that was suggested.  After that we'll try to test.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/4288/HighFidelity-Beta-PR11611-c79bf3bc042b8e45ba7af8f393a14851f715416b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/4288/HighFidelity-Beta-PR11611-c79bf3bc042b8e45ba7af8f393a14851f715416b.exe

**Comment by AndrewMeadows** [comment]
Ok this is ready to test.  Basically, if the performance isn't worse, and nothing is broken, we'll want to merge this.

**Comment by Nex-Pro** [comment]
T E S T I N G  win10

**Comment by Nex-Pro** [comment]
### Probable PASS on the test requirements. Slight performance gain.

***See inline notes re. content display issue.***

Test Plan

(1) Login to dev-welcome and find a spot off to the side that looks toward most of the objects in the domain.


Used this location:   hifi://dev-welcome/126.004,39.8939,-281.351/0,0.966926,0,0.255056

(2) Logout to save your login location. This location will be used to compare performance stats, so don't change it or the avatar's look-at direction for the initial comparison tests:

(3) Login to your known location using the latest dev-download interface client.

(4) Make sure these timing options are set as such:
Developer --> Timing --> Performance Timing --> Show Timing Details = enabled
Developer --> Timing --> Performance Timing --> Show Top Ten = disabled
Developer --> Timing --> Performance Timing --> Expand /update = enabled
Developer --> Timing --> Performance Timing --> Expand /physics = enabled

(5) Press / key to toggle the stats visible.

(6) Click the mouse on the stats to toggle the details visible.

(7) Pay attention to the stats value for: /idle/update/physics/harvestChanges.  

Wait for the time value to settle down. It will probably eventually reach a slow oscillation: pay attention to the max/min values of the oscillation. How much it oscillates is probably dependent on where you stand relative to the moving/animating objects in the domain. Try a position farther out from center, putting most of the objects in front of the avatar's view. **First run got 5.3 to 5.5*

(8) Logout, then back in to the same spot and repeat measure in (7) to make sure the approximate value is repeatable.
**Second run got 3.02 to 3.12**
**Third run got 2.95 to 3.10**
**Fourth run got 2.95 to 3.2  Typical 2.99**


(9) Using the same login location and orientation: repeat step (8) using the interface client from this PR.
--> The min/max times spent in the /idle/update/physics/harvestChanges context should be smaller when using this PR.

**First run in this PR got 2.55 to 3.02** 
**Second run in this PR got 2.22 to 3.12**
**Third run in this PR got 2.52 to 2.60**

This is about as consistent as I was able to observe.

(10) Walk around dev-welcome looking for objects that are animated (there's a flag blowing in the wind and other objects).
--> Verify animated models are still animating in this PR. **SUCCESS**
**NB: I had to reload content several times and restart interface at this point as I could only see the builder's grid.***

e26304-hifi-snap-by--on-2017-10-20_02-02-09

(11) Throw some dynamic objects around
--> verify dynamic objects still appear to move **SUCCESS**

(12) Find another avatar.
--> verify other avatars still render and update correctly **SUCCESS**
